### PR TITLE
Implement Special Linear (nguigs-hiring)

### PR DESCRIPTION
**Pull request by lpereira95**
Here it goes my implementation of special linear along with corresponding tests.

Looking forward for the feedback.

### CODE REVIEW

File: geomstats/geometry/special_linear.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+        Returns
+        -------
+        belongs : array-like, shape=[...,]
+            Boolean evaluating if point belongs to the manifold.
+        """
+        has_right_shape = self.lie_algebra.ambient_space.belongs(point)
+
+        if gs.all(has_right_shape):
+            # ???: det can be negative?
+            return gs.isclose(gs.abs(gs.linalg.det(point)), 1.0, atol=atol)
```
Comment (nguigs) [inline geomstats/geometry/special_linear.py:52]: No this space is restricted to positive determinant

File: geomstats/geometry/special_linear.py [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+        point : array-like, shape=[..., n, n]
+            Point in embedding manifold.
+
+        Returns
+        -------
+        projected : array-like, shape=[..., n, n]
+            Projected point.
+        """
+        scale_coeff = gs.power(gs.abs(gs.linalg.det(point)),
+                               1.0 / self.n)
```
Comment (nguigs) [inline geomstats/geometry/special_linear.py:73]: `algebra_utils.flip_determinant` can be used to make sure det > 0

File: geomstats/geometry/special_linear.py [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+        n_non_diag_half = (self.dim - n + 1) // 2
+
+        self.basis = gs.zeros((self.dim, n, n))
+
+        self.basis[dims[:n_non_diag_half], rows, cols] = 1.
+        self.basis[dims[n_non_diag_half:-n + 1], cols, rows] = 1.
+
+        self.basis[dims[-n + 1:], 0, 0] = -1
+        self.basis[dims[-n + 1:], diag_rows[1:], diag_cols[1:]] = 1
+
```
Comment (nguigs) [inline geomstats/geometry/special_linear.py:138]: Assignment won't work in tensorflow...

File: geomstats/geometry/special_linear.py [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+        Integer representing the shape of the matrices: n x n.
+    """
+
+    def __init__(self, n):
+        super(SpecialLinearLieAlgebra, self).__init__(
+            dim=n**2 - 1,
+            n=n,
+        )
+        # ???: not sure it is in the right place
+        self.ambient_space = GeneralLinear(n)
```
Comment (nguigs) [inline geomstats/geometry/special_linear.py:122]: Should be `Matrices(n)`

File: geomstats/geometry/special_linear.py [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+        point: array-like, shape=[..., n, n]
+            Point.
+
+        Returns
+        -------
+        point: array-like, shape=[..., n, n]
+            Projected point.
+        """
+        projected_point = gs.copy(point)
+        projected_point[..., 0, 0] = projected_point[..., 0, 0] - gs.trace(
```
Comment (nguigs) [inline geomstats/geometry/special_linear.py:208]: Assignment -> you can use the first element of the basis: `p -= trace * e0`

**Review by nguigs — CHANGES_REQUESTED** [review]
Great first contribution! Please avoid assignment to make it work with Tensorflow

File: geomstats/geometry/special_linear.py [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+
+        Returns
+        -------
+        projected : array-like, shape=[..., n, n]
+            Projected point.
+        """
+        scale_coeff = gs.power(gs.abs(gs.linalg.det(point)),
+                               1.0 / self.n)
+        aux_shape = [1] if len(gs.shape(point)) == 2 else [1, 1]
+        return gs.divide(point, gs.reshape(scale_coeff, (-1, *aux_shape)))
```
Comment (nguigs) [inline geomstats/geometry/special_linear.py:75]: I didn't know divide worked in that case (multiple arrays). There is an issue related to divide that you could look at: #636

File: geomstats/geometry/special_linear.py (hunk 1 above)
Comment (lpereira95) [reply geomstats/geometry/special_linear.py:52]: return gs.isclose(gs.linalg.det(point), 1.0, atol=atol)

File: geomstats/geometry/special_linear.py (hunk 2 above)
Comment (lpereira95) [reply geomstats/geometry/special_linear.py:73]: I guess this solves the problem (and avoids use of divide):

```
        proj_point = utils.flip_determinant(point, gs.linalg.det(point))
        scale_coeff = gs.power(gs.linalg.det(proj_point), 1. / self.n)
        return gs.einsum('...ij,...->...ij', proj_point, 1. / scale_coeff)
```

File: geomstats/geometry/special_linear.py (hunk 6 above)
Comment (lpereira95) [reply geomstats/geometry/special_linear.py:75]: At the end we can avoid divide and use `gs.einsum` as in other parts of the package.

File: geomstats/geometry/special_linear.py (hunk 5 above)
Comment (lpereira95) [reply geomstats/geometry/special_linear.py:208]: this solves the issue (after `Matrices.basis` implementation `gs.array_from_sparse` can be replaced):

```
trace = gs.trace(point, axis1=-2, axis2=-1)
# TODO: use ambient_matrix basis instead
sps_matrix = gs.array_from_sparse([(0, 0)], [1.0], (self.n, self.n))

return gs.copy(point) - gs.einsum('...,...ij->...ij', trace, sps_matrix)
```

**Comment by lpereira95** [comment]
I will close this now and open a new one, since it was open too long ago and tons of commits were done in the meantime.

**Pull request closed without merging**